packages/kbn-config/src/raw/read_config.ts

```diff
@@ -26,40 +26,73 @@ function replaceEnvVarRefs(val: string) {
   });
 }
 
-function merge(target: Record<string, any>, value: any, key?: string) {
-  if (isPlainObject(value) && Object.keys(value).length > 0) {
-    for (const [subKey, subVal] of Object.entries(value)) {
-      merge(target, subVal, key ? `${key}.${subKey}` : subKey);
+interface YamlEntry {
+  path: string[];
+  value: any;
+}
+
+const listEntries = (root: Record<string, any>): YamlEntry[] => {
+  const entries: YamlEntry[] = [];
+
+  const recursiveListEntries = (currentLevel: any, currentPath: string[]) => {
+    if (isPlainObject(currentLevel) && Object.keys(currentLevel).length > 0) {
+      for (const [subKey, subVal] of Object.entries(currentLevel)) {
+        const subKeySplits = splitKey(subKey);
+        recursiveListEntries(subVal, [...currentPath, ...subKeySplits]);
+      }
+    } else if (currentPath.length) {
+      entries.push({
+        path: currentPath,
+        value: processEntryValue(currentLevel),
+      });
     }
-  } else if (key !== undefined) {
-    set(target, key, recursiveReplaceEnvVar(value));
-  }
+  };
 
-  return target;
-}
+  recursiveListEntries(root, []);
+  return entries;
+};
 
-function recursiveReplaceEnvVar(value: any) {
+const mergeEntries = (entries: YamlEntry[]): Record<string, any> => {
+  const root = {};
+
+  const mergeEntry = (entry: YamlEntry) => {
+    set(root, entry.path, entry.value);
+  };
+  entries.map(mergeEntry);
+
+  return root;
+};
+
+function processEntryValue(value: any) {
   if (isPlainObject(value) || Array.isArray(value)) {
     for (const [subKey, subVal] of Object.entries(value)) {
-      set(value, subKey, recursiveReplaceEnvVar(subVal));
+      const unsplitKey = getUnsplittableKey(subKey);
+      if (unsplitKey) {
+        delete value[subKey];
+        set(value, [unsplitKey], processEntryValue(subVal));
+      } else {
+        set(value, subKey, processEntryValue(subVal));
+      }
     }
-  }
-  if (typeof value === 'string') {
+  } else if (typeof value === 'string') {
     return replaceEnvVarRefs(value);
   }
   return value;
 }
 
-/** @internal */
 export const getConfigFromFiles = (configFiles: readonly string[]) => {
-  let mergedYaml = {};
+  const yamlEntries: YamlEntry[] = [];
 
   for (const configFile of configFiles) {
     const yaml = readYaml(configFile);
     if (yaml !== null) {
-      mergedYaml = merge(mergedYaml, yaml);
+      yamlEntries.push(...listEntries(yaml));
     }
   }
 
-  return ensureDeepObject(mergedYaml);
+  for (const entry of yamlEntries) {
+    ensureValidObjectPath(entry.path);
+  }
+
+  return mergeEntries(yamlEntries);
```

Comment:
I kinda had to change the whole implementation of  `getConfigFromFiles` to add the feature.

FWIW, I tried a few approach, one trying to do this lower level in `ensureDeepObject`, but I couldn't find a way to do it without introducing a breaking change to config loading (mostly because of the fact that our config merge stops at the first array encountered by design), so in the end I was forced to do it there.

Not that it doesn't make sense anyway, I'm fine with the changes being done there.